File: cpp/middle/find_bottom_left_tree_value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int findBottomLeftValue(TreeNode* root) {
        if(root == NULL)
            return -1;
        int level = 1;
        int max_level = 1;
        int most_left_val = root->val;
        dfs_helper(root, level, max_level, most_left_val);
        return most_left_val;
    }
    
    void dfs_helper(TreeNode* root, int depth, int& max_depth, int& most_left_val)
    {
        if(!root)
            return;
        if(depth > max_depth)
        {
            most_left_val = root->val;
            max_depth = depth;
        }
        dfs_helper(root->left, depth+1, max_depth, most_left_val);
        dfs_helper(root->right, depth+1, max_depth, most_left_val);
    }
};

Conclusion:
an middle level algorithm issue; however, I think this one should be the easy level issue, dfs preorder is pretty good enough to solve this
algorithm issue, each time, we compare the max level has been reached with current level, if current level is larger than the max level has
been reached, then update the most_left_val, and each time we visit the left side first of all; thus, we will always get the most left value
in this algorithm.

Time Complexity: O(n)
Space Complexity: O(1)

File: python/middle/find_bottom_left_tree_value.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    most_left_val = 1
    max_depth = 1
    def findBottomLeftValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if(root == None):
            return 0
        depth = 1
        self.most_left_val = root.val
        def dfs(root, level):
            if(not root):
                return
            if(level > self.max_depth):
                self.most_left_val = root.val
                self.max_depth = level
            dfs(root.left, level+1)
            dfs(root.right, level+1)
        dfs(root, depth)
        return self.most_left_val
        
